camera_system/optimal_motion_searcher.py

```diff
@@ -0,0 +1,233 @@
+"""最適動作探索モジュール.
+
+走行体が指定されたノードに移動するための最適動作を探索する.
+@author: KakinokiKanta miyashita64
+"""
+
+import numpy as np
+import copy
+import operator
+from typing import List, Dict
+from game_area_info import GameAreaInfo
+from node import Node, NodeType
+from robot import Robot, Direction
+from coordinate import Coordinate
+from composite_game_motion import CompositeGameMotion
+from motion_converter_mock import MotionConverterMock  # ToDo: 動作変換が実装でき次第差し替える.
+
+
+class OptimalMotionSearcher:
+    """最適動作探索クラス."""
+
+    # 縦横移動にかかる最低限のコストの目安
+    __VERTICAL_COST = 0.5480
+    # 斜め移動にかかる最低限のコストの目安
+    __DIAGONAL_COST = 0.7840
+
+    @classmethod
+    def search(cls, start_robot: Robot, goal_node: Node) -> CompositeGameMotion:
+        """開始時の走行体から目標ノードに遷移するための最適動作を探索する.
+
+        Args:
+            start_robot: 開始時の走行体
+            goal_node:   目標ノード
+        Returns:
+            目標ノードに遷移するためのゲーム動作群: CompositeGameMotion
+        """
+        # 探索する走行体のハッシュ値を保持
+        open = [cls.robot_hash(start_robot)]
+        # 走行体のハッシュ値をキーに探索情報を保持
+        # ToDo: エッジを保持する
+        transition_table = {
+            open[0]: {"robot": copy.deepcopy(start_robot),                      # 走行体
+                      "motions": CompositeGameMotion(),                         # 開始時からの動作群
+                      "cost": cls.predict_cost(start_robot, goal_node.coord),   # 推定コスト
+                      "logs": [copy.deepcopy(start_robot)]}                     # 走行体の推移
+        }
+
+        # 探索対象がブロック設置動作か、ブロック取得動作か(True:設置, False:取得)
+        is_set_motion = goal_node.node_type != NodeType.BLOCK.value
+        # 設置動作探索時、ゴールノードにブロックがある場合
+        if is_set_motion and goal_node.block_id != -1:
+            print("A block already exists at the goal node.")
+            # 空のCompositeGameMotionを返す
+            return CompositeGameMotion()
+
+        # 最適動作を探索する
+        while transition_table[open[0]]["robot"].coord != goal_node.coord:
+            # openの先頭(推定コストが最小)を取り出す
+            min_cost_transition = transition_table[open.pop(0)]
+            # 遷移可能な走行体を取得する
+            next_robots = cls.next_robots(
+                min_cost_transition["robot"], goal_node.coord, is_set_motion)
+
+            # 遷移可能な走行体について探索する
+            for robot in next_robots:
+                # 走行体のハッシュ値を求める
+                hash = cls.robot_hash(robot)
+                # 開始時の走行体 -> 1ゲーム動作前の走行体 の動作群
+                motions = copy.deepcopy(min_cost_transition["motions"])
+                # 1ゲーム動作前の走行体 -> 探索対象の走行体 の動作
+                motion = MotionConverterMock.convert(min_cost_transition["robot"], robot)
```

Comment:
ここを見ただけで遷移してることを見てわかるように，第二引数はnextが入ってる方がいいと思う．ので，65行目は`for next_robot in next_robots`かなぁ．
できれば第一引数もcurrent_robotとかnow_robotとかの方がいいと思う．
`現在→次`って変数名の方が説明的でよくない？
だとすると`min_cost_transition`自体を変えるか，ここだけのために一時変数をよういするかだけどけ，まぁ納得いかなったら第一引数はいいや

あと，ここもgame_motionの方がいいと思う
```suggestion
                game_motion = MotionConverterMock.convert(current_robot, next_robot)
```